device/Microsoft.Azure.Devices.Client/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -454,26 +454,38 @@ public void OnConnected()
 
         public async void OnMessageReceived(Message message)
         {
-            if ((this.State & TransportState.Open) == TransportState.Open)
+            // Added Try-Catch to avoid unknown thread exception
+            // after running for more than 24 hours
+            try
             {
-                if (message.MqttTopicName.StartsWith(twinResponseTopicPrefix))
-                {
-                    twinResponseEvent(message);
-                }
-                else if (message.MqttTopicName.StartsWith(twinPatchTopicPrefix))
-                {
-                    await HandleIncomingTwinPatch(message).ConfigureAwait(false);
-                }
-                else if (message.MqttTopicName.StartsWith(methodPostTopicPrefix))
+                if ((this.State & TransportState.Open) == TransportState.Open)
                 {
-                    await HandleIncomingMethodPost(message).ConfigureAwait(false);
-                }
-                else
-                {
-                    this.messageQueue.Enqueue(message);
-                    this.receivingSemaphore.Release();
+                    if (message.MqttTopicName.StartsWith(twinResponseTopicPrefix))
+                    {
+                        twinResponseEvent(message);
+                    }
+                    else if (message.MqttTopicName.StartsWith(twinPatchTopicPrefix))
+                    {
+                        await HandleIncomingTwinPatch(message).ConfigureAwait(false);
+                    }
+                    else if (message.MqttTopicName.StartsWith(methodPostTopicPrefix))
+                    {
+                        await HandleIncomingMethodPost(message).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        this.messageQueue.Enqueue(message);
+                        this.receivingSemaphore.Release();
+                    }
                 }
             }
+            catch(Exception ex)
+            {
+                // This can still throw a fatal exception
```

Comment:
I'm not sure what you mean by this comment: "which defeats the reason ..." 
Applications should never catch exceptions that are fatal such as OutOfMemory or AccessViolation. In that case the application should immediately terminate, which is what is happening today. 

That said, the recommendation is that all expected exceptions (network, timeout) are caught in the DotNetty callbacks and properly handled (in this case by the OnError call which will, in turn, correctly switch the state machine and report back to the initial application call.

Please catch only exceptions that are not fatal so that the stack is preserved correctly. See https://github.com/Azure/azure-iot-sdk-csharp/pull/327/files#diff-95d848600b3ef1d9923119856d544fe5R538 on how to do